core/src/main/java/io/questdb/cutlass/pgwire/CircuitBreakerRegistry.java

```diff
@@ -0,0 +1,127 @@
+
+import io.questdb.cairo.CairoConfiguration;
+import io.questdb.cairo.CairoException;
+import io.questdb.cairo.sql.NetworkSqlExecutionCircuitBreaker;
+import io.questdb.mp.SimpleSpinLock;
+import io.questdb.std.*;
+
+import java.io.Closeable;
+
+// registry of circuit breakers utilized by pg wire contexts 
+// makes it possible to cancel queries executed in other connections/threads
+public class CircuitBreakerRegistry implements Closeable {
+
+    private final ObjList<NetworkSqlExecutionCircuitBreaker> circuitBreakers;
+    private final IntList freeIdx;
+
+    private final SimpleSpinLock lock;
+
+    private final Rnd random;
+
+    public CircuitBreakerRegistry(PGWireConfiguration configuration, CairoConfiguration cairoConfig) {
+        lock = new SimpleSpinLock();
+        int limit = configuration.getDispatcherConfiguration().getLimit();
+        circuitBreakers = new ObjList<>(limit);
+        freeIdx = new IntList(limit);
+        if (configuration.getRandom() != null) {
+            // use separate rnd to avoid changing existing pg test results 
+            random = new Rnd(configuration.getRandom().getSeed0(), configuration.getRandom().getSeed1());
+        } else {
+            random = new Rnd(
+                    cairoConfig.getNanosecondClock().getTicks(),
+                    cairoConfig.getMicrosecondClock().getTicks());
+        }
+
+        for (int i = 0; i < limit; i++) {
+            circuitBreakers.add(new NetworkSqlExecutionCircuitBreaker(configuration.getCircuitBreakerConfiguration(), MemoryTag.NATIVE_CB5));
+            freeIdx.add(i);
+        }
+    }
+
+    public int acquire() {
+        int idx;
+        int secret;
+        lock.lock();
+        try {
+            int size = freeIdx.size();
+            idx = freeIdx.getQuick(size - 1);
+            freeIdx.setPos(size - 1);
+            secret = random.nextInt();
+        } finally {
+            lock.unlock();
+        }
+
+        circuitBreakers.get(idx).setSecret(secret);
+        return idx;
+    }
+
+    public void cancel(int circuitBreakerIdx, int secret) {
+        if (circuitBreakers.size() < circuitBreakerIdx || circuitBreakerIdx < 0) {
+            throw CairoException.nonCritical().put("wrong circuit breaker idx [idx=").put(circuitBreakerIdx).put("]");
+        }
+
+        NetworkSqlExecutionCircuitBreaker cb = circuitBreakers.get(circuitBreakerIdx);
+        if (cb == null) {
+            throw CairoException.nonCritical().put("empty circuit breaker slot [idx=").put(circuitBreakerIdx).put("]");
+        }
+
+        if (cb.getSecret() != secret) {
+            throw CairoException.nonCritical().put("wrong circuit breaker secret [idx=").put(circuitBreakerIdx).put("]");
+        }
+
+        cb.cancel();
+    }
+
+    @Override
+    public void close() {
+        lock.lock();
+        try {
+            for (int i = 0, n = circuitBreakers.size(); i < n; i++) {
+                Misc.free(circuitBreakers.getQuick(i));
```

Comment:
Can you close the acquired instances here? 
I think it's better to set the acquired instances to a canceled state here and wait until the owning context closes them.

Reply:
Fixed in  fd45433e .